sdk/include/opentelemetry/sdk/metrics/state/temporal_metric_storage.h

```diff
@@ -0,0 +1,47 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+#ifndef ENABLE_METRICS_PREVIEW
+#  include "opentelemetry/sdk/metrics/state/attributes_hashmap.h"
+#  include "opentelemetry/sdk/metrics/state/metric_collector.h"
+
+#  include <memory>
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace metrics
+{
+
+struct LastReportedMetrics
+{
+  std::unique_ptr<AttributesHashMap> attributes_map;
+  opentelemetry::common::SystemTimestamp collection_ts;
+};
+
+class TemporalMetricStorage
+{
+public:
+  TemporalMetricStorage(InstrumentDescriptor instrument_descriptor);
+
+  bool buildMetrics(CollectorHandle *collector,
+                    nostd::span<std::shared_ptr<CollectorHandle>> collectors,
+                    opentelemetry::common::SystemTimestamp sdk_start_ts,
+                    opentelemetry::common::SystemTimestamp collection_ts,
+                    std::shared_ptr<AttributesHashMap> delta_metrics,
+                    nostd::function_ref<bool(MetricData)> callback) noexcept;
+
+private:
+  InstrumentDescriptor instrument_descriptor_;
+
+  // unreported metrics stash for all the collectors
+  std::unordered_map<CollectorHandle *, std::list<std::shared_ptr<AttributesHashMap>>>
```

Comment:
Why do we need `std::list` here?

Reply:
This code in TemporalMetricStorage is mostly same as in SyncmetricStorage for maintain unreported metrics and last reported metrics;
 https://github.com/open-telemetry/opentelemetry-cpp/blob/13ac5c30b6f8ae996d05f3c4084c58f952734ac6/sdk/include/opentelemetry/sdk/metrics/state/sync_metric_storage.h#L118

These structures are explained in the Sync Metric storage flow in design doc (https://docs.google.com/document/d/17PiDI91MQy07kkiSJ7lvlnogvMD3FV85g__WEackj1M/edit).

TemporalMetricStroage is supposed to be shared between Async and Sync storage. Once this PR is merged, I will raise another PR to use TemporalMetricStorage for Sync storage.

Reply:
Thanks for the clarification.